File: ports/go/src/core/tools/bcsgeometry.go
package tools

// BcsPoint defines an X/Y coordinate
type BcsPoint struct {
	X int
	Y int
}

func NewBcsPoint(x, y int) *BcsPoint {
	return &BcsPoint{X: x, Y: y}
}

// BcsSize defines width and height dimensions
type BcsSize struct {
	Width  int
	Height int
}

func NewBcsSize(w, h int) *BcsSize {
	return &BcsSize{Width: w, Height: h}
}

// BcsRect defines a 2D bounding box
type BcsRect struct {
	X      int
	Y      int
	Width  int
	Height int
}

func NewBcsRect(x, y, w, h int) *BcsRect {
	return &BcsRect{X: x, Y: y, Width: w, Height: h}
}

func (r *BcsRect) Contains(p *BcsPoint) bool {
	return p.X >= r.X && p.X <= r.X+r.Width &&
		p.Y >= r.Y && p.Y <= r.Y+r.Height
}

func (r *BcsRect) Intersects(other *BcsRect) bool {
	return r.X < other.X+other.Width && r.X+r.Width > other.X &&
		r.Y < other.Y+other.Height && r.Y+r.Height > other.Y
}

File: ports/rust/src/core/tools/bcsgeometry.rs
// BcsPoint defines an X/Y coordinate
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct BcsPoint {
    pub x: i32,
    pub y: i32,
}

impl BcsPoint {
    pub fn new(x: i32, y: i32) -> Self {
        BcsPoint { x, y }
    }
}

// BcsSize defines width and height dimensions
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct BcsSize {
    pub width: i32,
    pub height: i32,
}

impl BcsSize {
    pub fn new(width: i32, height: i32) -> Self {
        BcsSize { width, height }
    }
}

// BcsRect defines a 2D bounding box
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct BcsRect {
    pub x: i32,
    pub y: i32,
    pub width: i32,
    pub height: i32,
}

impl BcsRect {
    pub fn new(x: i32, y: i32, width: i32, height: i32) -> Self {
        BcsRect { x, y, width, height }
    }

    pub fn contains(&self, p: &BcsPoint) -> bool {
        p.x >= self.x && p.x <= self.x + self.width &&
        p.y >= self.y && p.y <= self.y + self.height
    }

    pub fn intersects(&self, other: &BcsRect) -> bool {
        self.x < other.x + other.width && self.x + self.width > other.x &&
        self.y < other.y + other.height && self.y + self.height > other.y
    }
}
